### DISCUSSION TITLE
A lot of useless log.Debug in Ocelot

### DISCUSSION THREAD

**Discussion by RaynaldM** [ThreeMammals/Ocelot#1736] [Ideas]
We're looking to speed up Ocelot a bit, and gain a few extra requests/second.
Using the benchmark, we ran a small test on [DownstreamRouteFinderMiddlewareBenchmarks](/search?q=repo%3AThreeMammals%2FOcelot%20DownstreamRouteFinderMiddlewareBenchmarks&type=code)
In [DownstreamRouteFinderMiddleware](/search?q=repo%3AThreeMammals%2FOcelot+DownstreamRouteFinderMiddleware&type=code), we put the `log.Debug` in `#if DEBUG` like this:
```csharp
       public async Task Invoke(HttpContext httpContext)
        {
            var upstreamUrlPath = httpContext.Request.Path.ToString();

            var upstreamQueryString = httpContext.Request.QueryString.ToString();

            var hostHeader = httpContext.Request.Headers["Host"].ToString();
            var upstreamHost = hostHeader.Contains(':')
                ? hostHeader.Split(':')[0]
                : hostHeader;

#if DEBUG
            Logger.LogDebug($"Upstream url path is {upstreamUrlPath}");
#endif

            var internalConfiguration = httpContext.Items.IInternalConfiguration();

            var provider = _factory.Get(internalConfiguration);

            var response = provider.Get(upstreamUrlPath, upstreamQueryString, httpContext.Request.Method, internalConfiguration, upstreamHost);

            if (response.IsError)
            {
                Logger.LogWarning($"{MiddlewareName} setting pipeline errors. IDownstreamRouteFinder returned {response.Errors.ToErrorString()}");

                httpContext.Items.UpsertErrors(response.Errors);
                return;
            }

#if DEBUG
            var downstreamPathTemplates = string.Join(", ", response.Data.Route.DownstreamRoute.Select(r => r.DownstreamPathTemplate.Value));
            Logger.LogDebug($"downstream templates are {downstreamPathTemplates}");
#endif

            // why set both of these on HttpContext
            httpContext.Items.UpsertTemplatePlaceholderNameAndValues(response.Data.TemplatePlaceholderNameAndValues);

            httpContext.Items.UpsertDownstreamRoute(response.Data);

            await _next.Invoke(httpContext);
        }
```
and the result was surprising:

With `log.Debug`
```
|   Method |     Mean |     Error |    StdDev |    StdErr |      Min |       Q1 |   Median |       Q3 |      Max |      Op/s | Ratio |  Gen 0 | Allocated |
|--------- |---------:|----------:|----------:|----------:|---------:|---------:|---------:|---------:|---------:|----------:|------:|-------:|----------:|
| Baseline | 1.232 us | 0.1523 us | 0.0236 us | 0.0118 us | 1.206 us | 1.217 us | 1.232 us | 1.248 us | 1.258 us | 811,506.6 |  1.00 | 0.3033 |      4 KB |
```
without
```
|   Method |     Mean |    Error |  StdDev |  StdErr |      Min |       Q1 |   Median |       Q3 |      Max |        Op/s | Ratio |  Gen 0 | Allocated |
|--------- |---------:|---------:|--------:|--------:|---------:|---------:|---------:|---------:|---------:|------------:|------:|-------:|----------:|
| Baseline | 890.6 ns | 28.16 ns | 4.36 ns | 2.18 ns | 884.2 ns | 889.7 ns | 892.4 ns | 893.3 ns | 893.4 ns | 1,122,855.0 |  1.00 | 0.2213 |      3 KB |
```
We gained around 300 request/s, i.e. a 38% improvement - not bad, eh?

I wonder if we couldn't do the same for others, or even all of them.

What do you think?

**Comment by raman-m** [comment]
Ray, when you said: "With LogDebug", did you mean in debug session enabled?

Yes, it is right direction...
Also, try to change logging severity in **launchSettings.json** and/or **appsettings.json** (aka `LogLevel` property) from **Debug** to **Error**/**Warning**, and you will get another surprising results! 🥳 😉

**Reply by RaynaldM** [reply]
Whether you're in a debug session or in release (which is the case in benchmark), if you call "logger.LogDebug(...)", it will really call the Logger, and it's the Logger, depending on the log level settings, that will log or not, but it's doing a job, and therefore consuming cpu.

I suggest enclosing all "LogDebugs" in "#if DEBUG" to prevent them from being executed in a release/production context.

**Reply by ggnaegi** [reply]
Hello, for me it's not obvious, in the benchmarks, what the logger minimum level is - because of Ocelot Logger design, we can't use some useful Logger extension methods directly - and if Debug logs are activated and if a debugger is attached...

**Reply by raman-m** [reply]
It seems we have a big problem with logger. As far as I know, Ocelot logger was  implemented as custom logger, and from your words, guys, I understood that this logger doesn't consider logging level from **appsettings.json**. This is the big design issue.
We have to remove custom implementation, and use standard dotnet logger of the box.

@RaynaldM  In Release build mode logger must ignore all debug records.
Moreover, logger must read the settings from app settings JSON and ignore all log records with severity lower than in the config. Microsoft default logger in dotnet provides such functionality, as you know.

**Reply by raman-m** [reply]
@RaynaldM Calling logger doesn't mean write log record actually. Logger should filter calls by severity. Calling a logger method takes a few nanoseconds. But real write operation of the record takes a few microseconds.

**Reply by raman-m** [reply]
> Hello, for me it's not obvious, in the benchmarks, what the logger minimum level is - because of Ocelot Logger design, we can't use some useful Logger extension methods directly - and if Debug logs are activated and if a debugger is attached...

@ggnaegi I don't want to say what gateway  :) but we have to check in repo what this team uses for logging. I believe they use standard Microsoft logging classes and extensions of the box. But we need to check it.
We don't need to reinvent the logging wheel. ☝️

**Reply by ggnaegi** [reply]
No, they don't reinvent the wheel, even for Rate Limiters...  Btw. The .net 7 Rate limiters are great and would fit the ocelot logic. But that's not part of the discussion.

**Reply by ggnaegi** [reply]
@RaynaldM @raman-m Maybe we can win some CPU cycles by avoiding the string transformations before calling Log.Debug...  Mmmh, no, but still...

**Reply by ggnaegi** [reply]
Probably a  set of standardized log factory methods with conditional attribute would greatly help...

**Reply by RaynaldM** [reply]
I never thought I'd start such a debate :smile:
We're talking about 28 "logDebugs" in Ocelot (and related projects).

@raman-m  calling the logger (which is standard) actually wastes a few nanoseconds (which add up to seconds) unnecessarily, since, normally in production, logDebugs shouldn't write anything, but the "LogDebug" function is actually called, it has to check that it shouldn't log anything and return to the caller, which is wasted CPU.
If you agree, we can easily make a PR where all logdebugs will be in "#if DEBUG" and won't even be compiled in Release mode.

**Reply by RaynaldM** [reply]
@ggnaegi I totally agree with you, we have to use all the standard resources at our disposal, and we'll have to talk about RateLimiter... one day 😄 .

**Reply by ggnaegi** [reply]
@RaynaldM I think a wrapper method would be the wise choice, then it's centralized? If you want, I can prepare the PR?

**Reply by raman-m** [reply]
> I never thought I'd start such a debate :smile:
> We're talking about 28 "logDebugs" in Ocelot (and related projects).
> 
> If you agree, we can easily make a PR where all logdebugs will be in "#if DEBUG" and won't even be compiled in Release mode.

@RaynaldM 
I totally agree. I have no objections to wrap `LogDebug` by preprocessor directive.

**Reply by raman-m** [reply]
> @RaynaldM I think a wrapper method would be the wise choice, then it's centralized? If you want, I can prepare the PR?

@ggnaegi 
That's perfect idea to encapsulate prepossessor directive inside of `LogDebug` implementation. I like it!
For string interpolation vs data we can use `Lazy<string>` value getter.

But `Func<string>` delegate parameter is better. Sounds good?
Instead of `Logger.LogDebug($"Upstream url path is {upstreamUrlPath}");` where interpolation & writing being performed,
...we could write like: `Logger.LogDebug(() => $"Upstream url path is {upstreamUrlPath}");` via anonymous delegate object to invoke it only in Debug mode. So, everything goes to encapsulation, and nothing executed until `DEBUG` mode.

**Reply by raman-m** [reply]
> @ggnaegi I totally agree with you, we have to use all the standard resources at our disposal, and we'll have to talk about RateLimiter... one day 😄 .

@RaynaldM Next design surprise by Rate Limiter? I have a strong feeling that someone at Microsoft just thieve Ocelot ideas... 😏 
It is OK to discuss. But open new idea discussion thread please!

**Reply by ggnaegi** [reply]
> @raman-m [Oct 17](#discussioncomment-7304820)

Yes that's exactly what I want to do, super optimization 😅

**Reply by RaynaldM** [reply]
why complicate things? these "logdebugs" have been around for a long time, and are useless for production purposes (unless you want to fill up a lot of useless log files), they're just for debugging, and I repeat, there are only 28 "LogDebug" in the entire source code.
![image](https://github.com/ThreeMammals/Ocelot/assets/606401/062d5430-53b3-4603-935f-5cfcdcc55a4c)


A simple Search/Replace and you're done.
If everyone agrees, of course 😄 .

**Comment by raman-m** [comment]
>@RaynaldM [wrote](#discussioncomment-7313856)
> A simple Search/Replace and you're done.
If everyone agrees, of course 😄

I agree not removing, but I do agree on wrapping by preprocessor directive.
I have no time for that. Maybe @ggnaegi will prepare a PR...

@RaynaldM and @ggnaegi  Have we reached a consensus? Will we turn this topic into an issue?

**Reply by ggnaegi** [reply]
@raman-m yes, I will be on it later today.

**Reply by raman-m** [reply]
@RaynaldM  Please, open issue by magic button [Create issue from discussion](../issues/new?created_from_discussion_number=1736) So, you'll be an author.

**Comment by raman-m** [comment]
The idea has been released by #1808
